### DISCUSSION TITLE
Tool specification input and output files

### DISCUSSION THREAD

**Discussion by Mastomaki** [Spine-project/Spine-Toolbox#2501] [Discussion]
In [https://spine-toolbox.readthedocs.io/en/latest/tool_specification_editor.html](https://spine-toolbox.readthedocs.io/en/latest/tool_specification_editor.html) is the idea that when you define some input and output files, they can be referenced inside the program, script so that file names are not hard coded?

**Comment by soininen** [comment]
The file names can be either hard coded or not. For example, you could have a script that *requires* (hard coded in the script) a file `input.dat` and *optionally* works with a bunch of files that are expected to have the `.csv` extension. In that case you would define

- `input.dat` as an *Input file*
- `*.csv` as *Optional input files*

The output files work similarly: you can hard code the entire file name or use wildcards for optional output files.

When specifying the Input and Output files in the Specification editor, Toolbox will copy the files to the Tool's work directory so they are available for the script in a known location. Note, that you can specify subdirectories for the files as well. These will be relative to the work directory.

These options expect some level of hard-coding: file names, or at least file extensions as well as relative locations to the work directory need to be known to the Tool script.

There is another, more general way to provide input files to the script that does not require any kind of hard coding: command line arguments. You can set them up in Tool's properties tab. For example, in the project below, I have set up a Data connection that provides input files for the workflow. The files are visible in Available resources list in Tool's properties and I have dragged and dropped them into the Tool arguments list.

![image](https://github.com/spine-tools/Spine-Toolbox/assets/19147159/5759cc2a-770a-4a93-9fc6-0796baafcf89)

Now my Python script can access the files by something like

```python
import sys
file_path1 = sys.argv[1]
file_path2 = sys.argv[2]
```

Of course, more serious scripts would use the [`argparse`](https://docs.python.org/3.12/library/argparse.html) module.

**Reply by Mastomaki** [reply]
Ok, thanks! I think the command line arguments approach suits me.

**Comment by ptsavol** [comment]
Maybe we should add @soininen's answer as-is to the User Guide?

**Reply by soininen** [reply]
Makes sense.

**Reply by Mastomaki** [reply]
I agree.

**Reply by ptsavol** [reply]
Updated the User Guide (commit 5cfaf9d)
